onnxruntime/core/providers/coreml/builders/impl/clip_op_builder.cc

```diff
@@ -0,0 +1,133 @@
+  // Both min and max will be inject in the layer, no need to add to the model
+  if (node.SinceVersion() >= 11) {
+    if (node.InputDefs().size() > 1)
+      model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
+
+    if (node.InputDefs().size() > 2)
+      model_builder.AddInitializerToSkip(node.InputDefs()[2]->Name());
+  }
+}
+
+Status ClipOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                            const Node& node,
+                                            const logging::Logger& logger) const {
+  const auto& node_name = node.Name();
+  const auto& input_name = node.InputDefs()[0]->Name();
+  const auto& output_name = node.OutputDefs()[0]->Name();
+  float min, max;
+  ORT_RETURN_IF_NOT(GetClipMinMax(model_builder.GetInitializerTensors(), node, min, max, logger), "GetClipMinMax failed");
+
+  bool has_min = min != std::numeric_limits<float>::lowest();
+  bool has_max = max != std::numeric_limits<float>::max();
+
+  if (!has_min && !has_max) {
+    // Clip without min/max is an identity node
+    // In CoreML we don't have identity, use ActivationLinear instead
+    std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+    layer->mutable_activation()->mutable_linear()->set_alpha(1.0f);
+    *layer->mutable_input()->Add() = input_name;
+    *layer->mutable_output()->Add() = output_name;
+
+    model_builder.AddLayer(std::move(layer));
+  } else {
+    // The implementation of clip(min, max) is done by
+    // 1. max(input, min) is handled by
+    //    min_output = threshold(input, min)
+    // 2. min(min_output, max) is handled by
+    //    output = -1 * (threshold(-min_output, -max))
+
+    // Now we have at least one or min or max is not default value
+    // Clipping max will need take the output of clipping min, or the node input, if min value is default
+    // If max value is default, the output of clipping min will be the output of the node
+    std::string min_output_name = output_name;
+    if (has_max) {
+      min_output_name = has_min
+                            ? model_builder.GetUniqueName(node_name + "min_output")
+                            : input_name;
+    }
+
+    // Handle min first
+    if (has_min) {
+      std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> min_layer = CreateNNLayer(node);
+      if (min == 0.0f) {  // If min is 0. then this min will be handled by relu
+        min_layer->mutable_activation()->mutable_relu();
+      } else {  // otherwise, min will be handled by unary->threshold
+        min_layer->mutable_unary()->set_alpha(min);
+        min_layer->mutable_unary()->set_type(COREML_SPEC::UnaryFunctionLayerParams::THRESHOLD);
+      }
+
+      *min_layer->mutable_input()->Add() = input_name;
+      *min_layer->mutable_output()->Add() = min_output_name;
+      model_builder.AddLayer(std::move(min_layer));
+    }
+
+    // Max is handled by -1 * (threshold (-min_output), -max))
```

Comment:
Also just curious, why do we need to apply the negative  `-1*` here, directly use `(threshold(min_output), max) will cause what issues?

Reply:
`threshold` is basically a `Max` operator
What we want is `Min(input, max)`, to use `threshold` you need to turn it into `-(Max(-input, -max))`

Reply:
Ah, I checked NeuralNetwork.proto, threshold is indeed a max operator.., Thanks for the explanation.